[TTAHUB-240] Include goals and objectives in CSV download

thewatermethod opened the pull request:
## Description of change
Goals and objectives are now included in CSV download (albeit in a clumsy fashion, until a more permanent solution is arrived at).

## How to test
Download report data for reports with mismatched numbers of goals. Reports should have empty columns where appropriate and data where appropriate.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-240


## Checklists

### Every PR

<!-- Add details to each completed item -->
- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [n/a] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

AdamAdHocTeam commented on src/lib/transform.js:
```diff
@@ -150,6 +152,24 @@ const arTransformers = [
   'lastSaved',
 ];
 
+function validateCsvRows(csvRows) {
```
The only concern I have is the ordering of the goal objectives as we discussed (minor).

If I enter obj 1, obj 2, obj 3 I get out obj 3, obj 2, obj 1 but there might not be a better way right now.

jasalisbury replied:
Can we rename this function to help explain it is adding missing columns for rows without all columns? Might also warrant a comment explaining the problem (I know I'll forget about it in a month or two!)

AdamAdHocTeam reviewed (approved):
It seems like this is something they urgently need. The only concern voiced below is the order of the Goal Objectives. If this is not a issue I would say good for now.

kryswisnaskas commented on src/routes/activityReports/handlers.js:
```diff
@@ -40,8 +40,9 @@ const logContext = {
+    validateCsvRows(csvRows),
     {
       header: true,
       quoted: true,
```
can we add destructuring on line 421 in this file to enable filtering
Suggestion:
Change `{ readRegions, limit: null },`
to `{ ...readRegions, limit: null },`

kryswisnaskas reviewed (changes requested):
Very good!

There are a few things I noticed.

- not related to this PR, but currently the filtering is not honored correctly for the export (see suggestion to fix)
- reports for non-grantees don't have goals, but the current code tries to access them. I am seeing: `2021-07-19T20:52:25.094Z AUDIT error:   SERVICE:ACTIVITY_REPORTS - UNEXPECTED ERROR - TypeError: Cannot read property 'name' of null` for non-grantees
- could we reconcile the naming and order of fields outlined in CSV fields_7.7.21  ?
( looks like the export will have goals and objectives columns appended at the end. Perhaps that's ok with Christine/Patrice/Damon, but maybe we should move all goals and objectives to the end? )

thewatermethod commented:
@kryswisnaskas 

I included your code suggestion and addressed the other issues (hopefully)

>     * reports for non-grantees don't have goals, but the current code tries to access them. I am seeing: `2021-07-19T20:52:25.094Z AUDIT error:   SERVICE:ACTIVITY_REPORTS - UNEXPECTED ERROR - TypeError: Cannot read property 'name' of null` for non-grantees
> 

I added some validation which seemed to resolve this on my end.

>     * could we reconcile the naming and order of fields outlined in CSV fields_7.7.21  ?
>       ( looks like the export will have extra goals and objectives columns appended at the end. Perhaps that's ok with Christine/Patrice/Damon, but maybe we should move all goals and objectives to the end? )

I went through and mapped the fields to the way they were presented in that spreadsheet. I also moved the goals and objectives to the end.

kryswisnaskas commented:
Thank you for implementing these changes! I verified that exporting a non-grantee report with one objective doesn't crash now. 👍 
We still get a crash when a non-grantee has multiple objectives. (I have an error msg if needed)

Couple of other minor changes - I noticed that for a non-grantee the objectives show up with a 0 as the first objective number. Could we change that to 1? 
Could we also capitalize the Goals and Objectives in columns to be more consistent with the rest of the columns?

Speaking of consistency, the .csv requirements file is consistent as far as capitalizing the first letter of columns, but less so with respect to the other words in the column names. We match most of the column names but not all. It would probably make sense to double check with Aricka or Lauren if what in the spreadsheet is exactly what we want and to match to minimize any future requests, e.g. is it `Virtual delivery type`  or `Virtual Delivery Type`? 😺

kryswisnaskas reviewed (changes requested):
see the comment

thewatermethod commented:
@kryswisnaskas I believe I have solved all the issues with the app crashing and the presentational issues as well 🤞

kryswisnaskas commented on src/lib/transform.js:
```diff
@@ -82,6 +85,11 @@ function makeGoalsAndObjectivesObject(objectiveRecords) {
+    if (!goalNum) {
+      goalNum = 1;
+    }
+
```
What do you think about making the same check for the objectives? Right now I am seeing Objective 1.0 for non-grantees. (unless the was done on purpose).

thewatermethod replied:
This is a good point, I was thinking it made sense the way I had it, however I realize that it will create other columns (like objective-1.0) that don't really make sense with the way the rest of them are named.

kryswisnaskas commented:
@thewatermethod Nice! The output looks great and it works! No crashes. 🎉 I still had one question, but after addressing it, I think this is good to go.

kryswisnaskas reviewed (approved):
🚀

kryswisnaskas commented on src/routes/activityReports/handlers.js:
```diff
@@ -51,6 +51,8 @@ async function sendActivityReportCSV(reports, res) {
   if (csvRows.length > 0) {
     const goalsAndObjectives = extractListOfGoalsAndObjectives(csvRows);
+    const user = await userById(userId);
+    const isAdmin = user.permissions.find((permission) => permission.scopeId === ADMIN);
```
This is ok for now, but maybe in the future we can take advantage of `src/policies/activityReport.js` to create a function e.g. `canPreview `or something similar.

kryswisnaskas reviewed (approved):
Looks good! I verified that admins only can see the new columns and that the goal status is not showing. 👍
